### PR TITLE
OOP school library: decorate a class

### PR DESCRIPTION
**Pull request by tomdan-ai**
# In This Milestone

#### Interface
- Create a class `Nameable`.
   - Implement a method called `correct_name` that will raise a `NotImplementedError`.

#### Turn your Person class to Nameable
- Make sure that your Person class inherits from Nameable
- Make sure that this class has a method `correct_name` implemented. It should simply return the name attribute.

#### Prepare base Decorator
- Make sure that it inherits from Nameable.
- In the constructor assign a nameable object from params to an instance variable.
- Implement the `correct_name` method that returns the result of the correct_name method of the `@nameable`.

#### Prepare CapitalizeDecorator and TrimmerDecorator
- For the CapitalizeDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that capitalizes the output of `@nameable.correct_name`.
- For the TrimmerDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that makes sure that the output of `@nameable.correct_name` has a maximum of 10 characters. If it's longer it should trim the word.

### CODE REVIEW

File: person.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 class Person
   attr_accessor :name, :age
```
Comment (ShadyShawkat) [inline person.rb]: - [ ] Please make sure that the `Person` inherits from the `Nameable` class, as per [this](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_decorator_v2.md#:~:text=Make%20sure%20that%20your%20Person%20class%20inherits%20from%20Nameable) requirement.

File: decorator.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+require_relative 'person'
+
+class Decorator < Person
```
Comment (ShadyShawkat) [inline decorator.rb]: - [ ] Please make sure that the `Decorator` inherits from the `Nameable` class, as per [this](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_decorator_v2.md#:~:text=Make%20sure%20that%20it%20inherits%20from%20Nameable.) requirement.

File: decorator.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  def initialize(person)
+    super(person.id, person.age, parent_permission: person.parent_permission, name: person.name)
+    @person = person
+  end
+
+  def correct_name
+    @person.correct_name
+  end
```
Comment (ShadyShawkat) [inline decorator.rb]: - [ ] Please make sure that the method returns the result of the `correct_name` method of the `@nameable` ,as per [this](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_decorator_v2.md#:~:text=Implement%20the%20correct_name%20method%20that%20returns%20the%20result%20of%20the%20correct_name%20method%20of%20the%20%40nameable.) requirement.
```ruby
@nameable.correct_name
```

**Review by ShadyShawkat — CHANGES_REQUESTED** [review]
## Hi @tomdan-ai, 

Good job so far! 👍🏻 
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 

Image

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb [hunk 4]
```diff
@@ -13,6 +16,10 @@ def can_use_services?
     of_age? || @parent_permission
   end
 
+  def correct_name
+    name
```
Comment (julie-ify) [inline person.rb]: - [ ] **[OPTIONAL]** This should be the instance variable `@name`, not `name`.

File: person.rb [hunk 5]
```diff
@@ -1,8 +1,11 @@
-class Person
+require_relative 'nameable'
+
+class Person < Nameable
   attr_accessor :name, :age
   attr_reader :id
 
   def initialize(id, age, parent_permission: true, name: 'Unknown')
+    super(id, age)
```
Comment (julie-ify) [inline person.rb]: - [ ] Kindly ensure you run your code and be sure it works. There was an error when testing your app due the super method here. Kindly remove this `id` and `age` from the super arguments. You should have something like this:

  ```
  super()
  ```

**Review by julie-ify — CHANGES_REQUESTED** [review]
Hi @tomdan-ai,

## Highlights!
- Trimmer class

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb (hunk 5 above)
Comment (tomdan-ai) [reply person.rb]: Thank You @julie-ify💯, I guess I was in a rush to submit, I forgot to test out my changes😊

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @tomdan-ai👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- No linter errors. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**